shared/src/main/scala/mlscript/ucs/Desugarer.scala

```diff
@@ -0,0 +1,605 @@
+          .getOrElseUpdate(scrutinee, MutMap.empty)
+          .getOrElseUpdate(fieldName, Var(freshName).withLoc(pattern.toLoc))
+        subPatterns += ((alias, pattern))
+        S(fieldName -> alias)
+    }.toList
+    subPatterns -> bindings
+  }
+
+  /**
+    * Desugar sub-patterns from fields to conditions.
+    *
+    * @param subPatterns a list of field name -> pattern term
+    * @param ctx the typing context
+    * @param aliasMap the field alias map
+    * @return desugared conditions representing the sub-patterns
+    */
+  private def destructSubPatterns(subPatterns: Iterable[Var -> Term])
+      (implicit ctx: Ctx, raise: Raise, aliasMap: FieldAliasMap, matchRootLoc: Opt[Loc]): Ls[Clause] = {
+    subPatterns.iterator.flatMap[Clause] {
+      case (scrutinee, subPattern) => destructPattern(scrutinee, subPattern)
+    }.toList
+  }
+
+  private val localizedScrutineeMap = MutMap.empty[Term, Var]
+
+  /**
+    * Create a `Scrutinee`. If the `term` is a simple expression (e.g. `Var` or
+    * `Lit`), we do not create a local alias. Otherwise, we create a local alias
+    * to avoid unnecessary computations.
+    *
+    * @param term the term in the local scrutinee position
+    * @param matchRootLoc the caller is expect to be in a match environment,
+    * this parameter indicates the location of the match root
+    */
+  def makeScrutinee(term: Term, isMultiLineMatch: Bool)(implicit matchRootLoc: Opt[Loc]): Scrutinee =
+    trace(s"Making a scrutinee for $term") {
+      val res = term match {
+        case _: Var => Scrutinee(N, term)
+        case _ =>
+          Scrutinee(
+            S(localizedScrutineeMap.getOrElseUpdate(term, {
+              Var(freshName).withLoc(term.toLoc)
+            })),
+            term,
+          )
+      }
+      res.isMultiLineMatch = isMultiLineMatch
+      res.matchRootLoc = matchRootLoc
+      res
+    }()
+
+  /**
+    * Destruct nested patterns to a list of simple condition with bindings.
+    *
+    * @param scrutinee the scrutinee of the pattern matching
+    * @param pattern the pattern we will destruct
+    * @param raise the `Raise` function
+    * @param aliasMap the field alias map
+    * @param matchRootLoc the location of the root of the pattern matching
+    * @param fragments fragment term that used to construct the given pattern.
+    *   It is used to tracking locations.
+    * @param isMultiLineMatch whether the scrutinee is in multi-line pattern match.
+    *   For example,
+    *   ```
+    *   if x is
+    *     A then "x is A!"
+    *     B then "x is B"
+    *   ```
+    *   is multi-line pattern match. Whereas `if x is Foo then 1 else 0` is not.
+    * @return a list of simple condition with bindings. This method does not
+    * return `ConjunctedCondition` because conditions built from nested patterns
+    * do not contain interleaved let bindings.
+    */
+  private def destructPattern
+      (scrutinee: Term, pattern: Term, isMultiLineMatch: Bool = false)
```

Comment:
Remove unnecessary default value. Defaults are error-prone so they should be reserved for easy-to-use APIs, not for core logic.

Reply:
Removed.